Zend/zend_compile.c

```diff
@@ -5255,53 +5271,51 @@ void zend_compile_match(znode *result, zend_ast *ast) /* {{{ */
 			Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
 		}
 		opnum_match = opline - CG(active_op_array)->opcodes;
-	}
-
-	uint32_t cond_count = 0;
-	uint32_t *jmpnz_opnums = safe_emalloc(sizeof(uint32_t), num_conds, 0);
-	for (uint32_t i = 0; i < arms->children; ++i) {
-		zend_ast *arm_ast = arms->child[i];
+	} else {
+		jmpnz_opnums = safe_emalloc(sizeof(uint32_t), num_conds, 0);
+		int cond_count = 0;
+		for (uint32_t i = 0; i < arms->children; ++i) {
+			zend_ast *arm_ast = arms->child[i];
 
-		if (!arm_ast->child[0]) {
-			if (has_default_arm) {
-				CG(zend_lineno) = arm_ast->lineno;
-				zend_error_noreturn(E_COMPILE_ERROR,
-					"Match expressions may only contain one default arm");
+			if (!arm_ast->child[0]) {
+				continue;
 			}
-			has_default_arm = 1;
-			continue;
-		}
 
-		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
-		for (uint32_t j = 0; j < conds->children; j++) {
-			zend_ast *cond_ast = conds->child[j];
+			zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+			for (uint32_t j = 0; j < conds->children; j++) {
+				zend_ast *cond_ast = conds->child[j];
 
-			znode cond_node;
-			zend_compile_expr(&cond_node, cond_ast);
+				znode cond_node;
+				zend_compile_expr(&cond_node, cond_ast);
 
-			if (expr_node.op_type == IS_CONST
-				&& Z_TYPE(expr_node.u.constant) == IS_FALSE) {
-				jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPZ, &cond_node, 0);
-			} else if (expr_node.op_type == IS_CONST
-				&& Z_TYPE(expr_node.u.constant) == IS_TRUE) {
-				jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPNZ, &cond_node, 0);
-			} else {
-				zend_op *opline = zend_emit_op(NULL, ZEND_IS_IDENTICAL, &expr_node, &cond_node);
-				SET_NODE(opline->result, &case_node);
-				if (opline->op1_type == IS_CONST) {
-					Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
+				if (expr_node.op_type == IS_CONST
+					&& Z_TYPE(expr_node.u.constant) == IS_FALSE) {
+					jmpnz_opnums[cond_count] = zend_emit_cond_jump(ZEND_JMPZ, &cond_node, 0);
```

Comment:
I missed this earlier: This looks like it was for `case true` and `case false` in **switch statements**.

JMPZ will jump if the argument is truthy. It would be correct for a switch (`if ($x == true)` behaves exactly the same way as `if ($x)` for all values), but is wrong for `match` statements. (e.g. `match($obj) { true => 'value' }`)

Similarly, JMPNZ jumps if the argument is falsey

So just always emit the ZEND_IS_IDENTICAL, and remove the if/else if

Reply:
Good catch, I didn't think about that. I'll add a test and remove this code.